File: tools/ez_tools/shellbags_parser.py
import os
import pandas as pd
from utils.discovery import find_artifact_files, load_csv_with_progress
from collector.collector import add_rows
from utils.logger import print_and_log
from utils.summary import track_summary

def process_shellbags(ez_dir: str, batch_size: int, base_dir: str):
    artifact_name = "Shellbags"
    print_and_log(f"[{artifact_name}] Scanning for relevant CSVs under: {ez_dir}")

    shellbag_files = find_artifact_files(ez_dir, base_dir, artifact_name)

    if not shellbag_files:
        print_and_log(f"[{artifact_name}] No Shellbags files found.")
        return

    date_columns = [
        ("LastWriteTime", "Last Write"),
        ("FirstInteracted", "First Interacted"),
        ("LastInteracted", "Last Interacted")
    ]

    for file_path in shellbag_files:
        print_and_log(f"[{artifact_name}] Processing: {file_path}")
        total_rows = 0
        try:
            for df in load_csv_with_progress(file_path, batch_size, artifact_name=artifact_name):
                timeline_data = []

                for _, row in df.iterrows():
                    for col, label in date_columns:
                        timestamp = row.get(col, "")
                        dt = pd.to_datetime(timestamp, utc=True, errors="coerce")
                        if pd.isnull(dt):
                            continue
                        dt_str = dt.isoformat().replace("+00:00", "Z")

                        timeline_row = {
                            "DateTime": dt_str,
                            "TimestampInfo": label,
                            "ArtifactName": artifact_name,
                            "Tool": "EZ Tools",
                            "Description": "File & Folder Access",
                            "DataPath": row.get("AbsolutePath", ""),
                            "DataDetails": row.get("Value", ""),
                            "EvidencePath": os.path.relpath(file_path, base_dir) if base_dir else file_path
                        }

                        timeline_data.append(timeline_row)

                total_rows += len(timeline_data)
                add_rows(timeline_data)
                track_summary("EZ Tools", artifact_name, len(timeline_data))
        except Exception as e:
            print_and_log(f"[{artifact_name}] Failed to parse {file_path}: {e}")
            continue

        print_and_log(f"[✓] Parsed {total_rows} timeline rows from: {file_path}")

File: Tools/EZTools/ShellbagsParser.cs
// Tools/EZTools/ShellbagsParser.cs
using CsvHelper;
using CsvHelper.Configuration;
using ForensicTimeliner.CLI;
using ForensicTimeliner.Interfaces;
using ForensicTimeliner.Models;
using ForensicTimeliner.Utils;
using System.Globalization;

namespace ForensicTimeliner.Tools.EZTools;

public class ShellbagsParser : IArtifactParser
{
    public List<TimelineRow> Parse(string inputDir, string baseDir, ArtifactDefinition artifact, ParsedArgs args)
    {
        var rows = new List<TimelineRow>();

        Logger.PrintAndLog($"[>] - [{artifact.Artifact}] Scanning for relevant CSVs under: [{inputDir}]", "SCAN");

        var files = Discovery.FindArtifactFiles(inputDir, baseDir, artifact.Artifact);
        if (!files.Any())
        {
            Logger.PrintAndLog($"[!] - [{artifact.Artifact}] No matching files found in: {inputDir}", "WARN");
            return rows;
        }

        foreach (var file in files)
        {
            Logger.PrintAndLog($"[+] - [{artifact.Artifact}] Processing: {Path.GetRelativePath(baseDir, file)}", "PROCESS");
            int timelineCount = 0;

            try
            {
                using var reader = new StreamReader(file);
                using var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    HeaderValidated = null,
                    MissingFieldFound = null
                });

                var records = csv.GetRecords<dynamic>();

                foreach (var record in records)
                {
                    var dict = (IDictionary<string, object>)record;

                    var dateFields = new Dictionary<string, string>
                    {
                        { "LastWriteTime", "Last Write" },
                        { "FirstInteracted", "First Interacted" },
                        { "LastInteracted", "Last Interacted" }
                    };

                    foreach (var pair in dateFields)
                    {
                        if (!dict.ContainsKey(pair.Key)) continue;

                        string rawTs = dict[pair.Key]?.ToString() ?? "";
                        if (string.IsNullOrWhiteSpace(rawTs)) continue;
                        if (!DateTime.TryParse(rawTs, out DateTime dt)) continue;

                        string dtStr = dt.ToString("o").Replace("+00:00", "Z");

                        string absPath = dict.TryGetValue("AbsolutePath", out var ap) ? ap?.ToString() ?? "" : "";
                        string val = dict.TryGetValue("Value", out var valObj) ? valObj?.ToString() ?? "" : "";

                        rows.Add(new TimelineRow
                        {
                            DateTime = dtStr,
                            TimestampInfo = pair.Value,
                            ArtifactName = artifact.Artifact,
                            Tool = artifact.Tool,
                            Description = artifact.Description,
                            DataPath = absPath,
                            DataDetails = val,
                            EvidencePath = Path.GetRelativePath(baseDir, file)
                        });

                        timelineCount++;
                    }
                }

                Logger.PrintAndLog($"[✓] - [{artifact.Artifact}] Parsed {timelineCount} timeline rows from: {Path.GetFileName(file)}", "SUCCESS");
                LoggerSummary.TrackSummary(artifact.Tool, artifact.Artifact, timelineCount);
            }
            catch (Exception ex)
            {
                Logger.PrintAndLog($"[{artifact.Artifact}] Failed to parse {file}: {ex.Message}", "ERROR");
            }
        }

        return rows;
    }
}
